From a4996bd52f95f8c3577d3d45baaa0a3ea2795ea8 Mon Sep 17 00:00:00 2001
From: chewitt <christianshewitt@gmail.com>
Date: Thu, 12 Mar 2020 15:50:13 +0000
Subject: [PATCH 146/148] HACK: drm/panfrost: add compatible for bifrost GPUs

---
 drivers/gpu/drm/panfrost/panfrost_drv.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/gpu/drm/panfrost/panfrost_drv.c b/drivers/gpu/drm/panfrost/panfrost_drv.c
index b7a618db3ee2..30a19b73ea24 100644
--- a/drivers/gpu/drm/panfrost/panfrost_drv.c
+++ b/drivers/gpu/drm/panfrost/panfrost_drv.c
@@ -656,6 +656,7 @@ static int panfrost_remove(struct platform_device *pdev)
 }
 
 static const struct of_device_id dt_match[] = {
+	{ .compatible = "arm,mali-bifrost" },
 	{ .compatible = "arm,mali-t604" },
 	{ .compatible = "arm,mali-t624" },
 	{ .compatible = "arm,mali-t628" },
-- 
2.17.1

